Haskell/Haskell.sublime-syntax

```diff
@@ -1,407 +1,1388 @@
+      push: record-body
+
+  record-body:
+    - meta_scope: meta.block.haskell
+    - include: block-end
+    - include: expressions
+    - include: else-pop
+
+###[ BRACKETS ]################################################################
+
+  lists:
+    # 3.7 Lists
+    # https://www.haskell.org/onlinereport/haskell2010/haskellch3.html#x8-340003.7
+    - match: \[
+      scope: punctuation.section.sequence.begin.haskell
+      push: list-body
+
+  list-end:
+    - match: \]
+      scope: punctuation.section.sequence.end.haskell
+      pop: 1
+
+  list-body:
+    - meta_scope: meta.sequence.list.haskell
+    - include: list-end
+    - include: records
+    - include: list-keywords
+    - include: expressions
+    - include: else-pop
+
+  list-keywords:
+    # 6.2.7. Generalised (SQL-like) List Comprehensions
+    # https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/generalised_list_comprehensions.html
+    - match: (?:group|by|using){{break}}
+      scope: keyword.control.list.haskell
+
+  quasi-quotes:
+    # https://wiki.haskell.org/Quasiquotation
+    - match: (\[)([\w'']+)(\|)
+      captures:
+        1: punctuation.section.quasi-quotes.begin.haskell
+        2: variable.function.quasi-quoter.haskell
+        3: punctuation.section.quasi-quotes.haskell
+      push: quasi-quotes-body
+
+  quasi-quotes-body:
+    - meta_scope: meta.quasi-quotes.haskell
+    - meta_content_scope: meta.string.haskell string.unquoted.haskell
+    - match: \|\]
+      scope: punctuation.section.quasi-quotes.end.haskell
+      pop: 1
+
+###[ PARENTHESES ]#############################################################
+
+  parens:
+    # 3.8 Tuples
+    # https://www.haskell.org/onlinereport/haskell2010/haskellch3.html#x8-360003.8
+    - include: empty-tuples
+    - include: unboxed-tuples
+    - include: prefix-parens-operators
+    - match: (?=\()
+      branch_point: parens
+      branch:
+        - group
+        - tuple
+
+  group:
+    - match: \(
+      scope: punctuation.section.group.begin.haskell
+      set: group-body
+
+  group-end:
+    - match: \)
+      scope: punctuation.section.group.end.haskell
+      pop: 1
+
+  group-body:
+    - meta_scope: meta.group.haskell
+    - match: ',|:(?!{{symbol}})'
+      fail: parens
+    - include: group-end
+    - include: records
+    - include: expressions
+    - include: else-pop
+
+  tuple:
+    - match: \(
+      scope: punctuation.section.sequence.begin.haskell
+      set: tuple-body
+
+  tuple-end:
+    - match: \)
+      scope: punctuation.section.sequence.end.haskell
+      pop: 1
+
+  tuple-body:
+    - meta_scope: meta.sequence.tuple.haskell
+    - include: tuple-end
+    - include: records
+    - include: expressions
+    - include: else-pop
+
+  empty-tuples:
+    # 3.9 Unit Expressions and Parenthesized Expressions
+    # https://www.haskell.org/onlinereport/haskell2010/haskellch3.html#x8-380003.9
+    - match: (\()\s*(\))
+      scope: meta.sequence.tuple.haskell
```

Comment:
We talked in https://github.com/sublimehq/Packages/pull/2387 whether an empty tuple is supposed to be a language constant where they have special meaning (notably in functional languages), but I while we seemed to have general agreement on this, the PRs haven't been merged yet. The description in the linked reference also supports this.

Reply:
I haven't followed that PR.

Reply:
When scoping `()` as constant, we should apply the same patterns to all the other constructors as well, I guess?

Token  | Descr  | Scope  |
----|----|----
`()` | unit expression | `constant.language.unit`
`(,)`, `(,,)`, `(,,,)`, `(,…,)` | tuple constructors | `constant.language.tuple`
`[]`   | list constructors | `constant.language.list`
`(->)` | function constructor | `constant.language.function`

see: https://www.haskell.org/onlinereport/haskell2010/haskellch6.html#x13-1200006.1.3

Do those really have the same meaning everywhere?

Reply:
And how about range tuples such as `(..)`?

Reply:
Tuple constructors are either a type or a function, not a constant.

I don't understand what you mean with the "function constructor". `->` is a syntactic token afaict.

The *empty* list constructor I agree should also be a constant, then.

Reply:
> Tuple constructors are either a type or a function, not a constant.

I understood the docs so that `()` is equal to `(,)`. Both denote an empty tuple. So if we decide to scope the first constant, shouldn't the others have the same scope, too then? 

Or other way round, why is `()` a constant. Maybe a type or constructor, too?

All other `(<symbol>)` constructs are scoped `meta.prefix` as I found it strange to stack `variable.function keyword.operator` to somehow denote their functional nature while maintaining operator highlighting.

> I don't understand what you mean with the "function constructor"

Unfortunatelly I don't recall where, but I copied some code from a Haskell documentation page which used `(->)` while docs said it to be a function constructor.  That's what the following test cases are copied from:

https://github.com/sublimehq/Packages/blob/f16e3fc50c812e722e38afa1adeec9df95fb0d89/Haskell/tests/syntax_test_haskell.hs#L1628-L1633

But I may have misunderstood those docs or might have been confused in other ways.

Reply:
> I understood the docs so that `()` is equal to `(,)`

`(,)` creates a two-tuple of its arguments and can be curried like any normal function. (Similarly, `(,,)` creates a 3-tuple.)

```haskell
Prelude> (,) 1 2 == (1, 2)
True
Prelude> with_1 = (,) 1
Prelude> with_1 2
(1,2)
```

> Or other way round, why is () a constant. Maybe a type or constructor, too?

Honestly, I don't know what a "unix expression" refers to and whether `()` is part of the typing system somehow. From gut feeling, I say it's a constant. Maybe @mitranim knows?

> That's what the following test cases are copied from:

I just get a syntax error (and not a "variable not in scope") when trying that:
```
Prelude> (->)
<interactive>:30:2: error: parse error on input ‘->’
```

I also don't recall ever seeing that, but I'm not 100% sure whether it might somehow still be legal. I'm not a Haskell expert, I just find it fun.

Reply:
With color schemes in mind, I think this indeed needs a bit more consideration when giving another scope than `punctuation` to `()`. Usually, almost all symbol and punctuation characters are scoped either `punctuation` or `keyword.operator`. One unsatisfying, but recently improved exception to this are lambda function arrows, which currently have `keyword.declaration.function.arrow` (JS, C#, Scala) or `keyword.declaration.function.anonymous` (D & Ruby) with the latest updates.

I'm not against scoping things like `()` as `constant` and have them colored by default, but I think there should at least be a way for color schemes to target these situations if they want to use a single color for all kind of symbol and punctuation characters. While the suggestions here with a `unit`, `list` etc. sub-scope would probably allow to distinguish them from other (word-based) constants, it's still annoying for color scheme authors to search for all such exceptions and inconsistencies between different syntaxes.

Btw, with `consant.language.list` here for `[]`, wouldn't the same apply e.g. for [Python](https://github.com/sublimehq/Packages/blob/366741f52ce8aaa5255bf24fa5bb00a043fff9b7/Python/syntax_test_python.py#L1217)?

Regarding the word vs. symbol operator problem mentioned in another comment above, most of it is discussed at #2090, but unfortunately there has not really been made progress. There was a suggestion from wbond to add a `meta.operator.word` scope to situations where an operator is a word. With these additional examples here, it would be great to have a general solution for such exceptions. Perhaps add something like `meta.symbol` on top of the scope for symbols/punctuations when they are not scoped as `punctuation` or `keyword.operator`, like this example here and lambda function arrows. And then maybe only introduce only `meta.word` on top of operator keywords to have them symmetric.

Reply:
> but I think there should at least be a way for color schemes to target these situations if they want to use a single color for all kind of symbol and punctuation characters

Would `constant.language punctuation` suffice or be too special cased?

> Btw, with consant.language.list here for [], wouldn't the same apply e.g. for Python?

No, because `[]` is a list literal and those are mutable and not constant. However, it would be valid for the empty tuple, since that is always the same.

```
>>> t = ()
>>> t is ()
<string>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
True
```

Reply:
I don't like the root of this `constants` discussion, to be honest. 

> With color schemes in mind, I think this indeed needs a bit more consideration when giving another scope than `punctuation` to `()`. Usually, almost all symbol and punctuation characters are scoped either `punctuation` or `keyword.operator`

That's exectly the reason for the original decision to use `meta.sequence punctuation` or `meta.group punctuation` for all kinds of such tokens, no matter whether they may be considered constant. That's also the reason for scoping arrows as `keyword.operator` rather than `keyword.other` which was the scope before this PR.

While it may help to see those empty `()` as constant I find it uncommon and odd compared to other languages such as python which don't (and shouldn't). I'd therefore be more than happy to revert the last commit.

> One unsatisfying, but recently improved exception to this are lambda function arrows

That's why I refuce to add another one by using `keyword.declaration` for `::` which denote type expressions. 

> And then maybe introduce only `meta.word` on top of operator keywords to have them symmetric.

We added additional meta scopes here and there, but scoping "each token" with several (espeially meta) scopes ... don't we find better solutions?

Reply:
I don't have strong opinions on scoping `()` or `[]` as constants in languages where they are frequently used as such, but I do recon this being quite useful for Rust, for example, where it's used as "the empty value", and it seems that Lisp people liked it as well. 

Python is arguably a different case, because it represents an empty tuple there, which is still "empty" but not the conventional "empty value", which would be `None`. To me, this differentiation makes sense and would mean that we also shouldn't scope the empty list in Haskell (`[]`) as constant, since it's an empty list and not "empty value". 
Are we confident enough to scope these kind of expressions as `constant.language.empty`? They exist in quite a few languages.

Reply:
Personally I would cast a vote against scoping `() [] {}` and such as constants, for fuzzy intuitive reasons that are hard to verbalize. But that's just one subjective opinion. Some of the reasons have already been voiced, such as color jitter when typing, or "opt in" vs "opt out" for color scheme support. Another is the issue of complexity: adding more code and special cases for no clear gain, while also having opposing voices. Erring on the side of simplicity rather than complexity is always a solid choice, in my opinion.

Reply:
> Would constant.language punctuation suffice or be too special cased?

That would work, however it could make the `constant.language` part ineffective for highlighting in some color schemes, because only the topmost colored scope "wins". Therefore it raises the question which of it is more important and should be on top; the `constant.language` or the `punctuation` property (i.e. why not use the order `punctuation constant.language`)? We had the same issue for logical operator keywords in e.g. Python, where both `keyword.operator.logical` and `keyword.operator.word` apply, but these are orthogonal properties. One argument in favor of `constant.language` in https://github.com/sublimehq/Packages/pull/2387#issuecomment-657299850 was
> that you might accidentally leave a stray () somewhere and it would be difficult to see themes which grey-out block scoping characters

So I assume it is desired here to have it colored as `constant.language` by default, but still have a way (e.g. with the uncolored meta-scope on top) for color schemes to opt-in for the "usual" punctuation coloring. Personally, I find the punctuation property more important, but I may be biased from the color scheme I use/maintain, and I don't use Haskell or Lisp-like languages myself.

> No, because [] is a list literal and those are mutable and not constant.

Okay, that makes sense. But then I wonder again if the priority really should be whether such objects are either mutable or immutable in a specific language, rather than have unified scopes for all tuples/arrays/lists etc. over all syntaxes. Most of the time `constant.language` is used for word-like constants like `true` and `null`, so that they stand out from variable identifiers, and for me it also seems uncommon to have `()` scoped as constant.

Python uses `meta.sequence.tuple.empty.python` in combination with the usual `punctuation` scopes for empty tuples, perhaps such a solution is sufficient here as well for users or color schemes that want to apply special highlighting to `()` and `[]`?

Reply:
Considdering all arguments of this discussion I decided to implement `meta.sequence.[list|tuple].empty`.

1. It is the most consistent solution with regards to highlighting. As the whole syntax is not designed to provide heavy/detailed semantic highlighting, it doesn't make sense to add such weird exceptions.
2. Stacking `constant` and `punctuaion` onto each other is avoided and `constant` feels out of place anyway, because `()` is used as type and constructor.

I consider this the final solution here and resolve this part of the discussion.